client/index.js

```diff
@@ -65,10 +66,27 @@ async function doRender ({ Component, props, err }) {
     props = await loadGetInitialProps(Component, { err, pathname, query })
   }
 
+  if (Component === ErrorComponent) {
+    lastScroll = {
+      x: window.pageXOffset,
+      y: window.pageYOffset
+    }
+  }
+
   Component = Component || lastAppProps.Component
   props = props || lastAppProps.props
 
   const appProps = { Component, props, err, router, headManager }
-  lastAppProps = appProps
   ReactDOM.render(createElement(App, appProps), container)
+
+  if (lastScroll &&
+    Component !== ErrorComponent &&
+    lastAppProps.Component === ErrorComponent) {
```

Comment:
It basically makes sure that the `Component` being rendered is a page and not an `ErrorComponent`.  Also checks if the last rendered component was an error, since we don't want lastScroll preservation on route changes and normal HMR.